pkg/vacuum/vacuum.go

```diff
@@ -196,68 +224,160 @@ func (e *Engine) Run(ctx context.Context) {
 	}
 }
 
 // listChunks identifies chunks which need to be vacuumed
-func (e *Engine) listChunks(ctx context.Context, con *pgxpool.Conn) ([]string, error) {
+func (e *Engine) listChunks(ctx context.Context, con *pgxpool.Conn) ([]*chunk, error) {
 	rows, err := con.Query(ctx, sqlListChunks)
 	if err != nil {
 		return nil, err
 	}
 	defer rows.Close()
-	tables := make([]string, 0)
+	chunks := make([]*chunk, 0)
 	for rows.Next() {
-		var table string
-		err := rows.Scan(&table)
+		c := chunk{}
+		err := rows.Scan(&c.name, &c.autovacuumCount, &c.age)
 		if err != nil {
 			return nil, err
 		}
-		tables = append(tables, table)
+		chunks = append(chunks, &c)
+	}
+	return chunks, nil
+}
+
+func (e *Engine) getVacuumSettings(ctx context.Context, con *pgxpool.Conn) (vacuumFreezeMinAge, autovacuumFreezeMaxAge int64, err error) {
+	err = con.QueryRow(ctx, sqlGetVacuumSettings).Scan(&vacuumFreezeMinAge, &autovacuumFreezeMaxAge)
+	if err != nil {
+		return 0, 0, err
+	}
+	return
+}
+
+// interpolate figures out how many workers to used based on the oldest transaction id age of the
+// chunks in our batch. The age should be between vacuumFreezeMinAge and autovacuumFreezeMaxAge.
+// The closer the maxChunkAge is to the autovacuumFreezeMaxAge, the more workers we assign. If
+// maxChunkAge is == vacuumFreezeMinAge, then we just use 1 worker.
+//
+//	       vacuumFreezeMinAge
+//	        |              maxChunkAge
+//	        |               |       autovacuumFreezeMaxAge
+//	0-------|---------------X--------|------
+//	        1000            6789     10000
+//
+// In this example, maxChunkAge is 64.3% of the way between vacuumFreezeMinAge and
+// autovacuumFreezeMaxAge. So, we will use 64.3% of maxParallelism workers (rounded).
+// If maxParallelism is 5, then we will use 3 workers.
+func interpolate(vacuumFreezeMinAge, autovacuumFreezeMaxAge, maxChunkAge, maxParallelism float64) int {
+	if maxChunkAge <= vacuumFreezeMinAge {
+		// this "shouldn't happen" but just in case
+		return 1
+	} else if autovacuumFreezeMaxAge <= maxChunkAge {
+		return int(maxParallelism)
 	}
-	return tables, nil
+	percent := (maxChunkAge - vacuumFreezeMinAge) / (autovacuumFreezeMaxAge - vacuumFreezeMinAge)
+	workers := math.Round(maxParallelism * percent)
+	// bounds check. workers should be between 1 and maxParallelism
+	if workers > maxParallelism {
+		workers = maxParallelism
+	} else if workers < 1 {
+		workers = 1
+	}
+	return int(workers)
+}
+
+// calcNumWorkers calculates the number of workers to use to vacuum chunks. The closer the oldest chunk
+// age gets to autovacuumFreezeMaxAge, the more workers we will use
+func (e *Engine) calcNumWorkers(ctx context.Context, con *pgxpool.Conn, chunks []*chunk) (int, error) {
+	vacuumFreezeMinAge, autovacuumFreezeMaxAge, err := e.getVacuumSettings(ctx, con)
+	if err != nil {
+		log.Error("msg", "failed to get vacuum settings", "error", err)
+		return 0, err
+	}
+	var maxChunkAge int64 = 0
+	for _, c := range chunks {
```

Comment:
nit: if we added an explicit order by age (again, I know the view has one) in the chunk query, we could avoid this loop and take the 1st item

Reply:
ah. nice.